dimod/cyvariables.pyx

```diff
@@ -0,0 +1,278 @@
+
+        Args:
+            v (hashable, optional):
+                Add a new variable. If `None`, a new label will be generated.
+                The generated label will be the index of the new variable if
+                that index is available, otherwise it will be the lowest
+                available non-negative integer.
+
+            permissive (bool, optional, default=False):
+                If `False`, appending a variable that already exists will raise
+                a `ValueError`. If `True`, appending a variable that already
+                exists will not change the container.
+
+        Returns:
+            hashable: The label of the appended variable.
+
+        Raises:
+            ValueError: If the variable is present and `permissive` is
+            False.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        if v is None:
+            v = self._stop
+
+            if not self._is_range() and self.count(v):
+                v = 0
+                while self.count(v):
+                    v += 1
+
+        elif self.count(v):
+            if permissive:
+                return v
+            else:
+                raise ValueError('{!r} is already a variable'.format(v))
+
+        idx = self._stop
+
+        if idx != v:
+            self._label_to_index[v] = idx
+            self._index_to_label[idx] = v
+
+        self._stop += 1
+        return v
+
+    cpdef bint _is_range(self):
+        """Return whether the variables are currently labelled [0, n)."""
+        return not PyDict_Size(self._label_to_index)
+
+    cpdef object _extend(self, object iterable, bint permissive=False):
+        """Add new variables.
+
+        Args:
+            iterable (iterable[hashable], optional):
+                An iterable of hashable objects.
+
+            permissive (bool, optional, default=False):
+                If `False`, appending a variable that already exists will raise
+                a `ValueError`. If `True`, appending a variable that already
+                exists will not change the container.
+
+        Raises:
+            ValueError: If a variable is present and `permissive` is
+            False.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        # todo: performance improvements for range etc
+        for v in iterable:
+            self._append(v, permissive=permissive)
+
+    def _pop(self):
+        """Remove the last variable.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        if not self:
+            raise IndexError("Cannot pop when Variables is empty")
+
+        self._stop = idx = self._stop - 1
```

Comment:
Will it handle this:
```python
variables = Variables('x')
variables._pop()
variables._pop()
```
Maybe it's a cython thing, but I'm not sure why would `self` be falsey after line 2 above?

Reply:
Uses [`__len__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__) in that case.